### DISCUSSION TITLE
How to change primary key generation to api tokens

### DISCUSSION THREAD

**Discussion by satheler** [adonisjs/core#1553] [Help]
I am usually using uuid as the primary key in all tables and I want to define this in API tokens table.
Is there any way to change the primary key value from increment integer to uuid values generated?

**Comment by miqbalrr** [comment]
you only need the primary key as uuid in Tokens table or all?
why you change that? i thought the tokens table is executed by adonis system too

**Reply by satheler** [reply]
I already managed to change the tables that I want to use the uuid. But to standardize my entire project I would also like to change this setting in the Tokens table.
The problem is not with the migration part, but with the model, I would have to change the id to string, for example.

**Comment by thetutlage** [comment]
If you will change the primary key to `uuid` then it will increase the token size massively, since the id is encoded within the generated token for faster lookups.

Technically uuid and auto increment counters both should work. Its just that they will impact the overall size of the token

**Reply by satheler** [reply]
I've managed to change the user model (for example) so that the id is uuid, that's not the problem.

The size of the token is also not an issue for me.

The problem is in the model that manages the api_tokens table, is there any way to change that instead of the primary key it is a string instead of a number?

**Reply by thetutlage** [reply]
Which version of the framework you are using? `v4.1` or `v5`

**Reply by satheler** [reply]
`v5`

**Reply by thetutlage** [reply]
In v5, we do not create any model for the API tokens. Can you share the code that you are talking about?

**Reply by satheler** [reply]
Currently, this is the database builder to insert a api token
```insert into `api_tokens` (`created_at`, `expires_at`, `name`, `token`, `type`, `user_id`) values (?, ?, ?, ?, ?, ?)```

but I would like that
```insert into `api_tokens` (`id` ,`created_at`, `expires_at`, `name`, `token`, `type`, `user_id`) values (?, ?, ?, ?, ?, ?, ?)```

Note: where I can provide the id value instead of database auto increment. This value would a uuid, for example.

A suggestion for make that:
New options as optional in authConfig inside provider key (`config/auth.ts`)

```typescript
/*
|--------------------------------------------------------------------------
| UUID as Identifier Key Value
|--------------------------------------------------------------------------
|
| The UUID as Identifier Key Value use uuid generated value instead of auto increment provided by database
|
*/
uuidAsIdentifierKeyValue?: boolean
```

And make a instruction to change database migration
```table.increments('id').primary()``` to ```table.uuid('id').primary()```

**Reply by thetutlage** [reply]
Got it. I have to think a bit about a better way to allow this (no answers for now). Also why don't you rely on the database to generate the uuid for you? If using PostgreSQL it has extensions for doing same.
